zeroex/order_js_test.go

```diff
@@ -0,0 +1,447 @@
+					Spender: constants.GanacheAccount4,
+					Value:   one,
+				},
+			},
+			err: errors.New("unknown event kind: FooBarBazEvent"),
+		},
+	} {
+		// Convert the contract event to a JSValue, and then attempt to
+		// recover it from the stringified JSON.
+		eventObject := testCase.event.JSValue()
+		eventString := stringify(eventObject)
+		var eventJSON contractEventJSON
+		err := json.Unmarshal([]byte(eventString), &eventJSON)
+		require.NoError(t, err)
+		decodedEvent, err := unmarshalContractEvent(&eventJSON)
+		if testCase.err != nil {
+			require.Equal(t, testCase.err, err)
+		} else {
+			require.NoError(t, err)
+			require.Equal(t, testCase.event, *decodedEvent)
+		}
+	}
+}
+
+func TestOrderEventConversion(t *testing.T) {
+	order := SignedOrder{
+		Order: Order{
+			MakerAddress:          constants.GanacheAccount0,
+			TakerAddress:          constants.NullAddress,
+			SenderAddress:         constants.NullAddress,
+			FeeRecipientAddress:   constants.GanacheAccount4,
+			MakerAssetData:        common.FromHex("0xf47261b0000000000000000000000000c02aaa39b223fe8d0a0e5c4f27ead9083c756cc2"),
+			MakerAssetAmount:      big.NewInt(10000000),
+			MakerFeeAssetData:     common.FromHex("0xf47261b000000000000000000000000089d24a6b4ccb1b6faa2625fe562bdd9a23260359"),
+			MakerFee:              big.NewInt(10000000),
+			TakerAssetData:        common.FromHex("0xf47261b000000000000000000000000081228eA33D680B0F51271aBAb1105886eCd01C2c"),
+			TakerAssetAmount:      big.NewInt(10000000),
+			TakerFeeAssetData:     common.FromHex("0xf47261b000000000000000000000000089d24a6b4ccb1b6faa2625fe562bdd9a23260359"),
+			TakerFee:              big.NewInt(10000000),
+			ChainID:               big.NewInt(1337),
+			ExpirationTimeSeconds: big.NewInt(0),
+			Salt:                  big.NewInt(0),
+		},
+		Signature: common.Hex2Bytes("0x1befcf4b6b1da4d207067a4b06e9bfbf21f85e2b6644f3ecf3a15f009e484756f251e3e00e909447ce45a16c620d14920a9acf516d9f4fe45bc36c914be6c9ec2703"),
+	}
+	orderHash, err := order.ComputeOrderHash()
+	require.NoError(t, err)
+	order.ResetHash()
+	for _, event := range []OrderEvent{
+		OrderEvent{
+			Timestamp:                time.Now().UTC(),
+			OrderHash:                orderHash,
+			SignedOrder:              &order,
+			EndState:                 "ADDED",
+			FillableTakerAssetAmount: big.NewInt(10000000),
+			ContractEvents: []*ContractEvent{
+				&ContractEvent{
+					BlockHash: blockHash,
+					TxHash:    txHash,
+					TxIndex:   txIndex,
+					LogIndex:  logIndex,
+					IsRemoved: false,
+					Address:   constants.GanacheAccount2,
+					Kind:      "ERC20ApprovalEvent",
+					Parameters: decoder.ERC20ApprovalEvent{
+						Owner:   constants.GanacheAccount3,
+						Spender: constants.GanacheAccount4,
+						Value:   one,
+					},
+				},
+			},
+		},
+	} {
+		eventObject := event.JSValue()
+		eventString := stringify(eventObject)
+		recoveredEvent := &OrderEvent{}
+		err := recoveredEvent.UnmarshalJSON([]byte(eventString))
```

Comment:
I actually think these tests are pretty close to what we want to do. If possible I want to avoid using `UnmarshalJSON` here since it is a confounding factor that could cover up some bugs. For example, in the `UnmarshalJSON` method we might do some conversions or handle empty/null values differently than in the actual flow from JavaScript to Go.